sdk/identity/src/device_code_flow/mod.rs

```diff
@@ -90,76 +89,87 @@ impl<'a> DeviceCodePhaseOneResponse<'a> {
     /// This will continue until either success or error is returned.
     pub fn stream(
         &self,
-    ) -> impl futures::Stream<Item = azure_core::Result<DeviceCodeAuthorization>> + '_ {
+    ) -> Pin<Box<impl futures::Stream<Item = azure_core::Result<DeviceCodeAuthorization>> + '_>>
+    {
         #[derive(Debug, Clone, PartialEq, Eq)]
         enum NextState {
             Continue,
             Finish,
         }
 
-        unfold(NextState::Continue, move |state: NextState| async move {
-            match state {
-                NextState::Continue => {
-                    let url = &format!(
-                        "https://login.microsoftonline.com/{}/oauth2/v2.0/token",
-                        self.tenant_id,
-                    );
-
-                    // Throttle down as specified by Azure. This could be
-                    // smarter: we could calculate the elapsed time since the
-                    // last poll and wait only the delta.
-                    sleep(Duration::from_secs(self.interval)).await;
-
-                    let encoded = form_urlencoded::Serializer::new(String::new())
-                        .append_pair("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
-                        .append_pair("client_id", self.client_id.as_str())
-                        .append_pair("device_code", &self.device_code)
-                        .finish();
-
-                    let http_client = self.http_client.clone().unwrap();
-
-                    match post_form(http_client.clone(), url, encoded).await {
-                        Ok(rsp) => {
-                            let rsp_status = rsp.status();
-                            let rsp_body = match rsp.into_body().collect().await {
-                                Ok(b) => b,
-                                Err(e) => return Some((Err(e), NextState::Finish)),
-                            };
-                            if rsp_status.is_success() {
-                                match serde_json::from_slice::<DeviceCodeAuthorization>(&rsp_body) {
-                                    Ok(authorization) => {
-                                        Some((Ok(authorization), NextState::Finish))
+        Box::pin(unfold(
```

Comment:
So the returned `Iterator` will just run forever on any caller running `next()` (apart from one error case I see). Shouldn't we guard against that, or is that based into futures' `unfold`? I admit I'm only somewhat familiar with `itertools::unfold`.

Reply:
As it relates to this PR, the only change in this PR is wrapping existing call to `unfold` with `Box::pin()`.

That said, I believe this has a handful of cases where `next()` will stop.

1. If the service returns success, but the body can't be read
2. If the service returns success and the body doesn't parse as intended
3. If the service returns non-success and the body doesn't parse as intended
4. If the service returns non-success and the response doesn't include `authorization_pending`
5. If the request itself fails (below the HTTP request layer)

I think it would be reasonable to add additional loop termination (such as maximum time we'll poll the service), but given the iterating over the stream is up to the caller, is this something that should be done in the SDK?

Either way, I think it would be appropriate to make such a change in a future PR.